File: pirates-android/app/src/main/java/com/crescentflare/piratesgame/components/simpleviewlets/TextViewlet.kt
package com.crescentflare.piratesgame.components.simpleviewlets

import android.content.Context
import android.support.v4.content.ContextCompat
import android.util.TypedValue
import android.view.Gravity
import android.widget.TextView
import com.crescentflare.piratesgame.R
import com.crescentflare.piratesgame.components.styling.AppFonts
import com.crescentflare.piratesgame.components.utility.MarkdownGenerator
import com.crescentflare.piratesgame.components.utility.ViewletUtil
import com.crescentflare.simplemarkdownparser.SimpleMarkdownConverter
import com.crescentflare.unilayout.views.UniTextView
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil


/**
 * Basic view viewlet: a text view
 */
object TextViewlet {

    // --
    // Viewlet instance
    // --

    val viewlet: JsonInflatable = object : JsonInflatable {

        override fun create(context: Context): Any {
            return UniTextView(context)
        }

        override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
            if (obj is UniTextView) {
                // Text
                val maxLines = mapUtil.optionalInteger(attributes, "maxLines", 0)
                val defaultTextSize = obj.resources.getDimensionPixelSize(R.dimen.text)
                val markdownText = ViewletUtil.localizedString(
                    obj.context,
                    mapUtil.optionalString(attributes, "localizedMarkdownText", null),
                    mapUtil.optionalString(attributes, "markdownText", null)
                )
                if (markdownText != null) {
                    val textInverted = ContextCompat.getColor(obj.context, R.color.textInverted)
                    val noColorization = mapUtil.optionalColor(attributes, "textColor", 0) == textInverted
                    val markdownString = SimpleMarkdownConverter.toSpannable(markdownText, MarkdownGenerator(obj.context.applicationContext, noColorization))
                    obj.setText(markdownString, TextView.BufferType.SPANNABLE)
                } else {
                    obj.text = ViewletUtil.localizedString(
                        obj.context,
                        mapUtil.optionalString(attributes, "localizedText", null),
                        mapUtil.optionalString(attributes, "text", null)
                    )
                }
                obj.setTextSize(TypedValue.COMPLEX_UNIT_PX, mapUtil.optionalDimension(attributes, "textSize", defaultTextSize).toFloat())
                obj.maxLines = if (maxLines == 0) Integer.MAX_VALUE else maxLines
                obj.setTextColor(mapUtil.optionalColor(attributes,"textColor", ContextCompat.getColor(obj.context, R.color.text)))

                // Font
                obj.typeface = AppFonts.getTypeface(mapUtil.optionalString(attributes, "font", "normal"))

                // Text alignment
                val textAlignment = TextAlignment.fromString(mapUtil.optionalString(attributes, "textAlignment", ""))
                obj.gravity = textAlignment.toGravity()

                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(mapUtil, obj, attributes)
                return true
            }
            return false
        }

        override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
            return obj is UniTextView
        }

    }


    // --
    // Default view creation
    // --

    fun newTextView(context: Context): UniTextView {
        val textView = UniTextView(context)
        textView.typeface = AppFonts.getTypeface("normal")
        textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, context.resources.getDimensionPixelSize(R.dimen.text).toFloat())
        textView.setTextColor(ContextCompat.getColor(context, R.color.text))
        return textView
    }


    // --
    // Text alignment enum
    // --

    enum class TextAlignment(val value: String) {

        Left("left"),
        Center("center"),
        Right("right");

        fun toGravity(): Int {
            return when(this) {
                Center -> Gravity.CENTER
                Right -> Gravity.RIGHT
                else -> Gravity.LEFT
            }
        }

        companion object {

            fun fromString(string: String?): TextAlignment {
                for (enum in TextAlignment.values()) {
                    if (enum.value == string) {
                        return enum
                    }
                }
                return Left
            }

        }

    }

}

File: pirates-ios/PiratesGame/Components/SimpleViewlets/TextViewlet.swift
//
//  TextViewlet.swift
//  Basic view viewlet: a text view
//

import UIKit
import UniLayout
import JsonInflator
import SimpleMarkdownParser

enum TextAlignment: String {
    
    case left = "left"
    case center = "center"
    case right = "right"
    
}

class TextViewlet {
    
    // --
    // MARK: Viewlet instance
    // --
    
    class func viewlet() -> JsonInflatable {
        return ViewletClass()
    }
    
    private class ViewletClass: JsonInflatable {
        
        func create() -> Any {
            return UniTextView()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let textView = object as? UniTextView {
                // Apply text styling
                let fontSize = convUtil.asDimension(value: attributes["textSize"]) ?? AppDimensions.text
                let font = AppFonts.font(withName: convUtil.asString(value: attributes["font"]) ?? "unknown", ofSize: fontSize)
                textView.font = font
                textView.numberOfLines = convUtil.asInt(value: attributes["maxLines"]) ?? 0
                textView.attributedText = nil
                textView.textColor = convUtil.asColor(value: attributes["textColor"]) ?? AppColors.text
                
                // Apply text alignment
                let textAlignment = TextAlignment(rawValue: convUtil.asString(value: attributes["textAlignment"]) ?? "") ?? .left
                switch textAlignment {
                case .center:
                    textView.textAlignment = .center
                case .right:
                    textView.textAlignment = .right
                default:
                    textView.textAlignment = .left
                }
                
                // Apply text (markdown needs to be set after text alignment)
                if let localizedMarkdownText = convUtil.asString(value: attributes["localizedMarkdownText"]) {
                    textView.attributedText = SimpleMarkdownConverter.toAttributedString(defaultFont: font, markdownText: localizedMarkdownText.localized(), attributedStringGenerator: MarkdownGenerator(noColorization: textView.textColor == AppColors.textInverted))
                } else if let markdownText = convUtil.asString(value: attributes["markdownText"]) {
                    textView.attributedText = SimpleMarkdownConverter.toAttributedString(defaultFont: font, markdownText: markdownText, attributedStringGenerator: MarkdownGenerator(noColorization: textView.textColor == AppColors.textInverted))
                } else if let localizedText = convUtil.asString(value: attributes["localizedText"]) {
                    textView.text = localizedText.localized()
                } else {
                    textView.text = convUtil.asString(value: attributes["text"])
                }
                
                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(convUtil: convUtil, view: textView, attributes: attributes)
                return true
            }
            return false
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return object is UniTextView
        }
        
    }
    
    class func newTextView() -> UniTextView {
        let textView = UniTextView()
        textView.font = AppFonts.normal.font(ofSize: AppDimensions.text)
        textView.textColor = AppColors.text
        return textView
    }
    
}
